corehq/apps/reports/filters/controllers.py

```diff
@@ -118,6 +133,14 @@ def _get_users(self, query, start, size, include_inactive=False):
             accessible_location_ids = SQLLocation.active_objects.accessible_location_ids(
                 self.request.domain, self.request.couch_user)
             users = users.location(accessible_location_ids)
+        if IS_ICDS_ENVIRONMENT and self.request.couch_user.get_role(self.domain).name == CPMU_ROLE_NAME:
+            # filtering out users who are in test locations for CPMU
+            test_locations = find_test_locations(self.domain)
+            users = users.filter(
```

Comment:
I think this would even work simply as
`users = users.location(location_ids)` since location is already present as a built in filter in https://github.com/dimagi/commcare-hq/blob/7b2dfa13d6a72be77f7b24c07d581f94ed7de5f3/corehq/apps/es/users.py#L47

Additionally, does not it always need ids/id and not the objects? Either case, since we just need the location ids, why not just fetch and cache ids instead of complete objects?

Reply:
> users = users.location(location_ids) since location is already present as a built in filter in

Yep. `users.location` is a built-in filter but what we want is users who are NOT in these location ids.  
So I had to use filters and go this way.

> Additionally, does not it always need ids/id and not the objects? Either case, since we just need the location ids, why not just fetch and cache ids instead of complete objects?

Yes Yes. For that `find_test_locations` is cached for a duration of 7 days.

Reply:
Ah, I see. thanks

Ya but we are caching objects and not just ids. we just needs ids right?

Reply:
No. We are caching just the set of ids, not the whole objects.

Reply:
ah, then we should name it `test_location_ids`.

Reply:
We should 👍